bucketswatcher_http.go

```diff
@@ -0,0 +1,300 @@
+	Endpoints        []string
+	Authenticator    Authenticator
+}
+
+type BucketsWatcherHttpOptions struct {
+	Logger               *zap.Logger
+	AutoDisconnectPeriod time.Duration
+}
+
+type bucketsWatcherHttpState struct {
+	tlsConfig        *tls.Config
+	authenticator    Authenticator
+	httpRoundTripper http.RoundTripper
+
+	endpoints []string
+}
+
+type BucketsWatcherHttp struct {
+	stateLock sync.Mutex
+	state     *bucketsWatcherHttpState
+	logger    *zap.Logger
+	userAgent string
+	closed    bool
+
+	agents    map[string]*Agent
+	lock      sync.Mutex
+	makeAgent func(ctx context.Context, bucketName string) (*Agent, error)
+
+	watcher        *StreamWatcherHttp[[]bucketDescriptor]
+	watchersCancel func()
+}
+
+func NewBucketsWatcherHttp(cfg BucketsWatcherHttpConfig, opts BucketsWatcherHttpOptions) (*BucketsWatcherHttp, error) {
+	logger := loggerOrNop(opts.Logger)
+
+	watcher, err := NewStreamWatcherHttp[[]bucketDescriptor](&StreamWatcherHttpConfig{
+		HttpRoundTripper: cfg.HttpRoundTripper,
+		Endpoints:        cfg.Endpoints,
+		UserAgent:        cfg.UserAgent,
+		Authenticator:    cfg.Authenticator,
+	}, &StreamWatcherHttpOptions{
+		Logger:               logger.Named("buckets-stream"),
+		AutoDisconnectPeriod: opts.AutoDisconnectPeriod,
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	bw := &BucketsWatcherHttp{
+		state: &bucketsWatcherHttpState{
+			authenticator:    cfg.Authenticator,
+			httpRoundTripper: cfg.HttpRoundTripper,
+			endpoints:        cfg.Endpoints,
+		},
+
+		makeAgent: cfg.MakeAgent,
+		userAgent: cfg.UserAgent,
+		logger:    logger,
+		agents:    make(map[string]*Agent),
+		watcher:   watcher,
+	}
+
+	return bw, nil
+}
+
+func (w *BucketsWatcherHttp) Reconfigure(cfg *BucketsWatcherHttpReconfigureConfig) error {
+	w.stateLock.Lock()
+	w.state = &bucketsWatcherHttpState{
+		httpRoundTripper: cfg.HttpRoundTripper,
+		endpoints:        cfg.Endpoints,
+		authenticator:    cfg.Authenticator,
+	}
+	w.stateLock.Unlock()
+
+	w.watcher.Reconfigure(&StreamWatcherHttpReconfigureConfig{
+		HttpRoundTripper: cfg.HttpRoundTripper,
+		Endpoints:        cfg.Endpoints,
+		Authenticator:    cfg.Authenticator,
+	})
+
+	return nil
+}
+
+func (bw *BucketsWatcherHttp) GetAgent(ctx context.Context, bucketName string) (*Agent, error) {
+	bw.lock.Lock()
+	if bw.closed {
+		bw.lock.Unlock()
+		return nil, errors.New("buckets watcher closed")
+	}
+
+	agent, ok := bw.agents[bucketName]
+	if ok {
+		bw.lock.Unlock()
+		return agent, nil
+	}
+	bw.lock.Unlock()
+
+	bw.logger.Debug("Bucket unknown, attempting manual update", zap.String("name", bucketName))
+
+	buckets, err := bw.manuallyFetchBuckets(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	bw.handleBuckets(buckets)
+
+	// We still don't know the bucket at this point then we can say it doesn't exist.
+	bw.lock.Lock()
+	agent, ok = bw.agents[bucketName]
+	bw.lock.Unlock()
+	if !ok {
+		return nil, cbmgmtx.ErrBucketNotFound
+	}
+
+	return agent, nil
+}
+
+func (bw *BucketsWatcherHttp) Watch() {
```

Comment:
This is a bit of a departure of our usual watcher API, which I don't love but kind of needs to be how it is.

Reply:
How do you see it as a departure?  Why does it need to be different?

Reply:
Our watchers typically take a context which is used for closing the watcher. For this watcher that doesn't work though as the buckets watcher owns a set of agents which need to be closed and doing that should block the caller. If we were to close the agents on context cancel we'd have no way of blocking the caller.